core/shared/src/main/scala/fs2/internal/FreeC.scala

```diff
@@ -200,20 +204,36 @@ private[fs2] object FreeC {
   object ViewL {
 
     /** unrolled view of FreeC `bind` structure **/
-    final case class View[F[_], X, R](step: F[X], next: Result[X] => FreeC[F, R])
-        extends ViewL[F, R]
+    sealed abstract case class View[F[_], X, R](step: F[X]) extends ViewL[F, R] {
+      def next(r: Result[X]): FreeC[F, R]
+    }
+
+    private[ViewL] final class EvalView[F[_], R](step: F[R]) extends View[F, R, R](step) {
+      def next(r: Result[R]): FreeC[F, R] = r.asFreeC[F]
+    }
 
     private[fs2] def apply[F[_], R](free: FreeC[F, R]): ViewL[F, R] = mk(free)
 
     @tailrec
     private def mk[F[_], R](free: FreeC[F, R]): ViewL[F, R] =
       free match {
-        case Eval(fx) => View(fx, pureContinuation[F, R])
+        case e: Eval[F, R] => new EvalView[F, R](e.fr)
         case b: FreeC.Bind[F, y, R] =>
-          b.fx match {
-            case Result(r)  => mk(b.f(r))
-            case Eval(fr)   => ViewL.View(fr, b.f)
-            case Bind(w, g) => mk(Bind(w, (e: Result[Any]) => Bind(g(e), b.f)))
+          b.step match {
+            case Result(r) => mk(b.cont(r))
+            case Eval(fr) =>
+              new ViewL.View[F, y, R](fr) {
+                def next(r: Result[y]): FreeC[F, R] = b.cont(r)
+              }
+            case bb: FreeC.Bind[F, z, _] =>
+              val nb = new Bind[F, z, R](bb.step) {
+                def cont(zr: Result[z]): FreeC[F, R] =
+                  new Bind[F, y, R](bb.cont(zr)) { self =>
+                    override val delegate: Bind[F, y, R] = b.delegate
```

Comment:
@djspiewak @pchlupacek  I am guessing that this was the cause for the strange behaviour that I was seeing in the benchmarks. 

Because this `b.delegate` is inside a `def cont`, this line was creating a back-reference to the `b` object, which prevented the garbage collector from cleaning up that `b` object, and which may have forced the evaluation of this line to reload an object. 

That would explain that, when running those benchmarks on the PR branch, we were seeing less allocations per _op_ (memory efficient); but there were also less throughput (ops per second), and more calls to the garbage collector (and more time in GC), because the GC could collect fewer objects in each run. 

I am pushing a commit with the correction to this, and the results of rerunning the benchmarks.